File: src/main/java/uk/gov/justice/hmpps/prison/service/OffenderAddressService.java
package uk.gov.justice.hmpps.prison.service;

import jakarta.validation.constraints.NotNull;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import uk.gov.justice.hmpps.prison.api.model.AddressDto;
import uk.gov.justice.hmpps.prison.repository.jpa.repository.OffenderAddressRepository;
import uk.gov.justice.hmpps.prison.repository.jpa.repository.OffenderRepository;

import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class OffenderAddressService {

    private final OffenderRepository offenderRepository;
    private final OffenderAddressRepository offenderAddressRepository;

    public List<AddressDto> getAddressesByOffenderNo(@NotNull final String offenderNo) {
        final var offender = offenderRepository.findOffenderWithLatestBookingByNomsId(offenderNo)
            .orElseThrow(EntityNotFoundException.withMessage(String.format("No offender found for offender number %s\n", offenderNo)));
        final var addresses = offenderAddressRepository.findByOffenderId(offender.getRootOffender().getId());
        return AddressTransformer.translate(addresses);
    }
}

File: src/main/java/uk/gov/justice/hmpps/prison/service/OffenderAddressService.kt
package uk.gov.justice.hmpps.prison.service

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import uk.gov.justice.hmpps.prison.api.model.AddressDto
import uk.gov.justice.hmpps.prison.repository.jpa.repository.OffenderAddressRepository
import uk.gov.justice.hmpps.prison.repository.jpa.repository.OffenderRepository

@Service
@Transactional(readOnly = true)
class OffenderAddressService(
  private val offenderRepository: OffenderRepository,
  private val offenderAddressRepository: OffenderAddressRepository,
) {
  fun getAddressesByOffenderNo(offenderNo: String): List<AddressDto> {
    val offender = offenderRepository.findRootOffenderByNomsId(offenderNo).orElseThrow()
    val addresses = offenderAddressRepository.findByOffenderId(offender.id)
    return AddressTransformer.translate(addresses)
  }
}
